sympy/matrices/common.py

```diff
@@ -2218,37 +2218,39 @@ def __neg__(self):
         return self._eval_scalar_mul(-1)
 
     @call_highest_priority('__rpow__')
-    def __pow__(self, num):
+    def __pow__(self, exp):
         if self.rows != self.cols:
             raise NonSquareMatrixError()
         a = self
-        jordan_pow = getattr(a, '_matrix_pow_by_jordan_blocks', None)
-        num = sympify(num)
-        if num.is_Number and num % 1 == 0:
+        exp = sympify(exp)
+
+        if exp.is_Number and exp % 1 == 0:
             if a.rows == 1:
-                return a._new([[a[0]**num]])
-            if num == 0:
+                return a._new([[a[0]**exp]])
+            if exp == 0:
                 return self._new(self.rows, self.cols, lambda i, j: int(i == j))
-            if num < 0:
-                num = -num
+            if exp == 1:
+                return self
+            if exp < 0:
+                exp = -exp
                 a = a.inv()
             # When certain conditions are met,
             # Jordan block algorithm is faster than
             # computation by recursion.
-            elif a.rows == 2 and num > 100000 and jordan_pow is not None:
+            elif a.rows == 2 and exp > 100000:
                 try:
-                    return jordan_pow(num)
-                except MatrixError:
-                    pass
-            return a._eval_pow_by_recursion(num)
-        elif not num.is_Number and num.is_negative is None and a.det() == 0:
+                    return a._matrix_pow_by_jordan_blocks(exp)
+                except (MatrixError, AttributeError):
+                    pass # a may not have _matrix_pow_by_jordan_blocks?
+            return a._eval_pow_by_recursion(exp)
+
+        try:
+            return a._matrix_pow_by_jordan_blocks(exp)
+        except ValueError:
+            if exp.is_negative or exp.is_number:
+                raise
```

Comment:
I think here it should be mentioned which error is raised.

Reply:
Re-raises the error which it is handling, the ValueError

Reply:
Okay. I had some tests in my mind and will try them if something different happens from expected will comment here.

Reply:
Catching AttributeError makes me nervous because it can hide bugs. I prefer the getattr(obj, 'attrname', None) idiom. AttributeError should always be a programmer error - if you aren't sure the attribute exists then hasattr/getattr should be used.

Reply:
Good point, will change